Source/CarthageKit/Xcode.swift

```diff
@@ -1542,15 +1563,25 @@ private func UUIDsFromDwarfdump(URL: NSURL) -> SignalProducer<Set<NSUUID>, Carth
 		}
 }
 
-/// Returns the URL of a binary inside a given framework.
-private func binaryURL(frameworkURL: NSURL) -> Result<NSURL, CarthageError> {
-	let bundle = NSBundle(path: frameworkURL.path!)
+/// Returns the URL of a binary inside a given package.
+private func binaryURL(packageURL: NSURL) -> Result<NSURL, CarthageError> {
+	let bundle = NSBundle(path: packageURL.path!)
+	let packageType = (bundle?.objectForInfoDictionaryKey("CFBundlePackageType") as? String).flatMap(PackageType.init)
 
-	if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
-		return .Success(frameworkURL.URLByAppendingPathComponent(binaryName))
-	} else {
-		return .Failure(.ReadFailed(frameworkURL, nil))
+	switch packageType {
+	case .Framework?:
+		if let binaryName = bundle?.objectForInfoDictionaryKey("CFBundleExecutable") as? String {
+			return .Success(packageURL.URLByAppendingPathComponent(binaryName))
+		}
+	case .dSYM?:
+		if let binaryName = packageURL.URLByDeletingPathExtension?.URLByDeletingPathExtension?.lastPathComponent {
+			let binaryURL = packageURL.URLByAppendingPathComponent("Contents/Resources/DWARF/\(binaryName)")
+			return .Success(binaryURL)
+		}
+	default: break
```

Comment:
Haha, no worries. I'm still not entirely pleased with how it reads either, so I'm open to feedback on better ways to set this flow up!